### ISSUE TITLE
Optic ignores severity level for spectral rules

### ISSUE THREAD

**Issue by ksteinershopify** [opticdev/optic#1826]
**Describe the bug**

I am currently working on incorporating `optic` as our open-api spec file linter and running into a problem where `optic` seems to ignore the severity levels of spectral rules. I am able to turn off rules but if a rule is not `off` but has any other severity level, if the rule is triggered it always prints out an failure message. I am not sure if this is an intent behavior or a bug, but was not able to find anything in the documentation around this.

**To Reproduce**

Using the following simple `spec.json` file with one endpoint
```
{
	"components": {
		"examples": {},
		"headers": {},
		"parameters": {},
		"requestBodies": {},
		"responses": {},
		"schemas": {
			"EventsResponseItem": {
				"properties": {
					"id": {
						"type": "string"
					},
					"type": {
						"type": "string"
					},
					"time": {
						"type": "string"
					},
					"payload": {
						"$ref": "#/components/schemas/PayloadSchema"
					}
				},
				"required": [
					"id",
					"type",
					"time",
					"payload"
				],
				"type": "object",
				"additionalProperties": false
			},
			"PayloadSchema": {
				"properties": {},
				"type": "object",
				"description": "Construct a type with a set of properties K of type T"
			},
			"EventId": {
				"type": "string",
				"pattern": "^([\\w]{0,50})$"
			}
		}
	},
	"info": {
		"title": "Test API",
		"version": "2021-01",
		"description": "Test API",
		"license": {
			"name": "ISC"
		},
		"contact": {
			"name": "John Doe",
			"email": "support@foo.com"
		}
	},
	"openapi": "3.0.0",
	"paths": {
		"/api/2021-01/events/{eventId}": {
			"get": {
				"operationId": "GetEvent",
				"responses": {
					"200": {
						"description": "OK",
						"content": {
							"application/json": {
								"schema": {
									"$ref": "#/components/schemas/EventsResponseItem"
								}
							}
						}
					}
				},
				"description": "Get specific event",
				"tags": [
					"Events"
				],
				"parameters": [
					{
						"in": "path",
						"name": "eventId",
						"required": true,
						"schema": {
							"$ref": "#/components/schemas/EventId"
						}
					}
				]
			}
		}
	},
	"servers": [
		{
			"url": "https://foo.com",
			"description": "Foo"
		}
	]
}
```
And using an `optic.dev.yml` file with a simple `always` spectral rule
```
ruleset:
  - spectral:
      always:
        - always.yml
```
and `always.yml` being
```
extends: "spectral:oas"
rules:
  operation-tag-defined: warn
  oas3-unused-component: off
```
with `operation-tag-defined` rule severity set to `warn`, `optic` is failing when running
```
$ npx optic lint spec.json
```
with
```
Checks

   FAIL  GET /api/2021-01/events/{eventId}
    requirement rule: Spectral always rule
      x Error code: operation-tag-defined: Operation tags must be defined in global tags.
      at ...
```
even though the `operation-tag-defined` severity level is set to warn. I would expect this only to happen if severity is set to `error`.

**Expected behavior**

Instead of failing here, it should only print out a warning. 

**Details (please complete the following information):**
 - Optic version: 0.42.4
 - NodeJS version: v16.17.0

**Comment by acunniffe** [comment]
Hey @ksteinershopify, I'm Aidan, the creator of this project. Thanks for opening this issue and sharing a complete reproduction. 

Looking at our code, it looks like we're hardcoding the the severity of each spectral result as "error". We'll open a PR this week to fix this and make sure Optic handles the options properly. Look for updates here. 

Re: Exit codes
- Optic exits `1` if there are `> 0 errors`
- When warn and info level are being used, it will not exit `1` when `warning > 0` unless there is an error in the same batch of results

Is that what you are expecting? 

P.s. glad to see you are using `added` vs `always` rules already 👍 . Anything else I can help with?

**Comment by acunniffe** [comment]
Makes sense -- we'll make sure the output communicates the severity level correctly as well. 

Are you using the visual changlog? We'll make sure to update that as well. 
![new-changelog](https://user-images.githubusercontent.com/5900338/231552740-7e941bc9-9788-4372-8c09-98c83840619f.jpg)
date that as well

**Comment by ksteinershopify** [comment]
> Makes sense -- we'll make sure the output communicates the severity level correctly as well.

Thank you so much!
> Are you using the visual changlog?

Not right now, we are currently using the cli tool only.

**Comment by acunniffe** [comment]
@ksteinershopify @gabsong  -- we've released support for spectral severity in 0.42.9. Just update your optic version and give it a go. We will keep this open until you confirm it is working as expected. Thanks!

**Comment by gabsong** [comment]
Thanks @acunniffe! I just installed it and confirmed the severity reporting is working as expected. I wanted to share a few extra things that is neither a bug or feature request -- just consider it feedback on my experience.

I have a few screenshots to share in case they are helpful -- we're trying to see why there is a difference in the number of warnings between `spectral` (26) and `optic` (41), since I was expecting an equal count. Sharing here in case you want to look in parallel.

Here's the [rules](https://gist.github.com/gabsong/833450326f773a63de6d0835aeba8dcf) that should have applied.

- [optic output](https://gist.github.com/gabsong/bd8e7089bc5a0674a032d64deaf30922)
- [spectral output](https://gist.github.com/gabsong/06924b965a62809dd0b45e00a8e9c0a0)

And since our API is public (not fully operational but it's on an unauthenticated page) here's the spec file as well: https://gist.github.com/gabsong/554bda56d03a0b344809369a7dccd2b4

Also, another point which is just related to UX of the output:

1. I thought that the dot.notation on the object of the error from the spectral output is helpful to Cmd+F and look for keywords. For example, I can search `ShipmentResponse` and find all logs associated to that term.
2. I think a column layout makes the output slightly easier to parse.

Given the following output:
```
requirement rule [warn]: Spectral always rule
      ⚠ Error code: ibm-schema-description: Schemas should have a non-empty description
      at /Users/shopify/Repos/network-api/spectral/temp/swagger.2022-10.json:615:17634
``` 

615 = this is the line in the file, I get this.
17634 = I'm not sure what this number is. Maybe it means the nth character, but I'm not sure if it's helpful

**Comment by niclim** [comment]
Hey - just looking at the rule count differences it looks like the spectral rules are running against the `components`, where as optic is running against the dereferenced spec (similar to [this issue](https://github.com/opticdev/optic/issues/1832#issuecomment-1508369557)). 

As for the UX suggestions - thanks for sharing your feedback, agreed that we can make the output easier to parse and some of the numbers aren't as useful - I'll look into updating the output here!

**Comment by gabsong** [comment]
Thanks Nic! The dereferencing makes sense. Now that we know we won't be as confused :)

Thank you for putting out a great tool!

**Comment by acunniffe** [comment]
@gabsong -- you will also notice the count deviates from Spectral if you run some rules on: 
```
added: ...
always: ...
```
This is because spectral runs all the rules on the current spec and Optic might only run certain rules on things added to the spec.

**Comment by niclim** [comment]
Just released `0.42.11` which updates the rule results to include the JSON path as part of the rule results https://github.com/opticdev/optic/pull/1852

As for the numbers after the `sourcemap`, your correct in the numbering, we include those numbers so that integrated terminals (e.g. vscode) can be clicked and take you to the exact place in your editor.

I'm going to close out this issue since the original issue is fixed - but if you find anything else feel free to open another issue - thanks for the detailed issue!
